### ISSUE TITLE
cyclic control flow for loops

### ISSUE THREAD

**Issue by carljm** [astral-sh/ty#232]
Currently we don't even model control flow back edges in loops (because they will often lead to inference cycles). Once we have fixpoint iteration support in Salsa, we need to add these back edges and tests for them, including cases where we have to fallback to avoid runaway fixpoint iteration, e.g.:

```py
x = 0
for _ in range(y):
    x += 1
```

**Comment by sharkdp** [comment]
I've seen multiple real world examples where missing support for loop control flow leads to an incorrect reachability analysis for large blocks of code within loops, because we observe a type that is too narrow. For example:

```py
first_iteration = True

for x in xs:
    if not first_iteration:
        # this whole block is considered unreachable

    # more code

    first_iteration = False
```

**Comment by Gankra** [comment]
I just encountered the exact same problem as David:

https://play.ty.dev/9214d74c-f4de-46df-80d0-44b3262d1ff4

```py
some_list: list[int] = [0, 1]

first_part = True
for j in range(0, 10):
    if first_part:
        first_part = False
        continue
    # ty infers `val: Never` here!?
    # (many other results now infer to Unknown)
    val = some_list[0]
```

And because we don't respect `first_part: bool` there's no (non-tedious) way to tell ty to chill.

**Comment by sharkdp** [comment]
> And because we don't respect `first_part: bool` there's no (non-tedious) way to tell ty to chill.

Obviously not ideal, but a `cast` always helps as a workaround in these situations (`first_part = cast(bool, True)`).

**Comment by anstadnik** [comment]
```python
    i = 0
    while i < 1: # Here i is Literal[0]
        i += 1 # Here i is Literal[1]
        
    reveal_type(i) # Never
```

`ty` thinks that it's an endless loop

**Comment by sin-ack** [comment]
Similar example, where Ty is overeager on narrowing based on the value it sees initially:

https://play.ty.dev/d8968435-e88b-4414-bf14-fa7611d14c0c

Interestingly, setting `foo` to `1` will narrow the type to `Literal[1]` in both branches, rather than `Never` and `1` on either branch respectively.

**Comment by oconnor663** [comment]
@sin-ack I don't think this is a narrowing issue per se. The problem is that, without loop-back bindings, `ty` is reasoning that either the `if` _or_ the `else` can execute, but not both. Happily, this example is already working [on my branch](https://github.com/astral-sh/ruff/pull/22794):

```
info[revealed-type]: Revealed type
  --> x.py:13:21
   |
11 |             foo = 1
12 |     else:
13 |         reveal_type(foo)
   |                     ^^^ `Literal[1]`
   |
```
